pkg/impersonation/impersonation.go

```diff
@@ -323,3 +344,86 @@ func (i *Impersonator) waitForServiceAccount(sa *corev1.ServiceAccount) (*corev1
 	}
 	return ret, nil
 }
+
+func getUser(clusterContext *config.UserContext, userInfo user.Info, groupname string) (user.Info, error) {
+	u, err := clusterContext.Management.Management.Users("").Controller().Lister().Get("", userInfo.GetUID())
+	if err != nil {
+		return &user.DefaultInfo{}, err
+	}
+
+	groups := []string{"system:authenticated", "system:cattle:authenticated"}
+	if groupname != "" {
+		groups = append(groups, groupname)
+	}
+	attribs, err := clusterContext.Management.Management.UserAttributes("").Controller().Lister().Get("", userInfo.GetUID())
+	if err != nil && !apierrors.IsNotFound(err) {
+		return &user.DefaultInfo{}, err
+	}
+	if attribs != nil {
+		for _, gps := range attribs.GroupPrincipals {
+			for _, groupPrincipal := range gps.Items {
+				groups = insertSortedIfNotPresent(groupPrincipal.Name, groups)
+			}
+		}
+	}
+
+	extras := getUserExtras(userInfo)
+	extras["username"] = insertSortedIfNotPresent(u.Username, extras["username"])
+	for _, principalid := range u.PrincipalIDs {
+		extras["principalid"] = insertSortedIfNotPresent(principalid, extras["principalid"])
+	}
+
+	user := &user.DefaultInfo{
+		UID:    u.GetName(),
+		Name:   u.Username,
+		Groups: groups,
+		Extra:  extras,
+	}
+	return user, nil
+}
+
+func insertSortedIfNotPresent(item string, list []string) []string {
```

Comment:
I assume this is for inserting into an already sorted list. I don't think it's working as expected: https://play.golang.org/p/Yko9VTmAqMd

One of those cases where creating unit tests for this func would be really valuable to ensure it's working as expected.

Reply:
Thanks for catching this, I've simplified this using sort